<a href="https://colab.research.google.com/github/sitthatkaja/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. ช่วงเช้าหรือช่วงบ่ายมีเครื่องบินขาเข้ามากกว่ากัน

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [1]:
import pandas as pd

In [2]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os

In [6]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM21/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [7]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM21/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [8]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [11]:
gen_exams('623020541-2')

รหัสนักศึกษา 623020541-2 ทำข้อ 2 และ 3


# 2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์

In [15]:
merged_data = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน')
merged_data

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

# 3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด

**จากตาราง merged_data เราจะหาสายการบินมากที่สุดจากการ Group by สายการบิน โดยใช้ . count ในการนับ**

In [47]:
merged_data.groupby('สายการบิน').count()

วันในเดือน  วันในสัปดาห์  ...     รัฐ  ชื่อสนามบิน
สายการบิน                            ...                     
9E              80031         80031  ...   80031        80031
AA             289855        289855  ...  289855       289855
AS              68555         68555  ...   68555        68555
B6             121906        121906  ...  121906       121906
DL             381657        381657  ...  381657       381657
EV             157928        157928  ...  157928       157928
F9              35738         35738  ...   35738        35738
FL              92702         92702  ...   92702        92702
HA              17432         17432  ...   17432        17432
MQ             113212        113212  ...  113212       113212
OO             160164        160164  ...  160164       160164
UA             286418        286418  ...  286418       286418
US             233321        233321  ...  233321       233321
VX              34739         34739  ...   34739        34739
WN             575739        575739  ...  575739       575739
YV              52821         52821  ...   52821        52821

[16 rows x 10 columns]

**ผลจากกาการใช้คำสั่ง merged_data.groupby('สายการบิน').count() พบว่า**

**วันในเดือนนั้นจะนับทีละ 1 ครั้งนับไปเรื่อยๆจากคำสั่ง count หมายความว่าหากค่าวันเดือนในเดือนมาก สายการบินนั้นจะมีเที่ยวบินที่มากที่สุด**

***

**เราจะใช้ให้ค่า sort values เป็น วันในเดือน โดยให้ค่า ascending เป็น False เพื่อให้เรียงจากมากลงไปน้อย**

In [50]:
summary = merged_data.groupby('สายการบิน').count().sort_values(by=['วันในเดือน'],ascending=False)
summary

วันในเดือน  วันในสัปดาห์  ...     รัฐ  ชื่อสนามบิน
สายการบิน                            ...                     
WN             575739        575739  ...  575739       575739
DL             381657        381657  ...  381657       381657
AA             289855        289855  ...  289855       289855
UA             286418        286418  ...  286418       286418
US             233321        233321  ...  233321       233321
OO             160164        160164  ...  160164       160164
EV             157928        157928  ...  157928       157928
B6             121906        121906  ...  121906       121906
MQ             113212        113212  ...  113212       113212
FL              92702         92702  ...   92702        92702
9E              80031         80031  ...   80031        80031
AS              68555         68555  ...   68555        68555
YV              52821         52821  ...   52821        52821
F9              35738         35738  ...   35738        35738
VX              34739         34739  ...   34739        34739
HA              17432         17432  ...   17432        17432

[16 rows x 10 columns]

**จากผลที่ได้จะพบว่าสายการบิน WN มีเที่ยวบินที่มากสุดจำนวน 575739 ครั้งและสายการบิน HA มีเที่ยวบินที่น้อยที่สุดจำนวน 17432 ครั้ง สามารถสรุปออกมาได้ดังนี้**

In [52]:
max = summary.iloc[:1,:1]
max

วันในเดือน
สายการบิน            
WN             575739

สายการบิน WN มีเที่ยวบินที่มากสุดจำนวน 575739 ครั้ง

In [55]:
min = summary.iloc[15:16,:1]
min

วันในเดือน
สายการบิน            
HA              17432

สายการบิน HA มีเที่ยวบินที่น้อยที่สุดจำนวน 17432 ครั้ง 